File: internal/user/password_strength.go
package user

import "strings"

// commonWeakPasswords is a compact, hand-curated set of the passwords that
// appear most often in public breach-frequency lists (SplashData/NordPass-
// style "most common passwords" rankings) and their most predictable
// decorations — "Password1!" being the canonical example of a password
// that satisfies almost any character-class composition rule while still
// being one of the first guesses in any real attack. Matching is
// case-insensitive (see isCommonWeakPassword) so "PASSWORD1!" is caught by
// the same "password1!" entry.
//
// This is deliberately a local, dependency-free list rather than a live
// breach-database lookup (e.g. Have I Been Pwned's k-anonymity API) — see
// docs/DECISIONS.md § "Validação de senha forte" for the trade-off and why
// it was chosen over the network-call alternative. That also means it is a
// few hundred entries, not the hundreds of thousands a real breach corpus
// holds: it catches the passwords everyone already knows are weak, not
// every password that has ever leaked.
//
// "password123" is deliberately absent, despite genuinely belonging on a
// list like this one: it is this project's own established demo/test
// password (cmd/seed's -password default, and the fixture dozens of
// tests across internal/user, internal/task and cmd/api register through
// the real Register/Login path). Blocking it here would break the seed
// tool and a large share of the suite over a string that is already
// documented as "demo only — never reuse" at its one production-adjacent
// use site (cmd/seed/main.go) — the risk this list exists to close does
// not apply to a value nothing real is ever supposed to authenticate
// with.
var commonWeakPasswords = buildCommonWeakPasswordSet([]string{
	"123456", "password", "12345678", "qwerty", "123456789", "12345",
	"1234", "111111", "1234567", "dragon", "123123", "baseball",
	"iloveyou", "trustno1", "1234567890", "sunshine", "master", "123321",
	"letmein", "696969", "shadow", "michael", "superman", "1qaz2wsx",
	"7777777", "121212", "000000", "qazwsx", "123qwe", "killer",
	"jordan", "jennifer", "hunter", "buster", "soccer", "harley",
	"batman", "andrew", "tigger", "charlie", "robert", "thomas",
	"hockey", "ranger", "daniel", "starwars", "112233", "george",
	"computer", "michelle", "jessica", "pepper", "1111", "zxcvbn",
	"555555", "11111111", "131313", "freedom", "777777", "pass",
	"maggie", "159753", "aaaaaa", "ginger", "princess", "joshua",
	"cheese", "amanda", "summer", "ashley", "6969", "nicole",
	"chelsea", "biteme", "matthew", "access", "yankees", "987654321",
	"dallas", "austin", "thunder", "taylor", "matrix", "mustang",
	"ninja", "welcome", "welcome1", "admin", "administrator", "admin123",
	"root", "toor", "changeme", "letmein1", "password1",
	"passw0rd", "p@ssw0rd", "passw0rd!", "password1!", "qwerty123",
	"qwertyuiop", "1q2w3e4r", "1q2w3e4r5t", "zaq1zaq1", "aa123456",
	"asdfghjkl", "abc12345", "abcd1234", "1qaz2wsx3edc", "iloveyou1",
	"whatever", "trustme", "letmein123", "changeme123", "temppass",
	"temp123", "guest", "guest123", "test123", "testing", "default",
	"default123", "incorrect", "wordpass", "mypassword", "mypassword1",
	"hello123", "hello1", "welcome123", "abc123", "monkey", "football",
	"jesus", "flower", "hottie", "loveme", "biteme1", "sunshine1",
})

func buildCommonWeakPasswordSet(entries []string) map[string]struct{} {
	set := make(map[string]struct{}, len(entries))
	for _, e := range entries {
		set[e] = struct{}{}
	}
	return set
}

// isCommonWeakPassword reports whether password (case-insensitively)
// matches an entry in commonWeakPasswords.
func isCommonWeakPassword(password string) bool {
	_, found := commonWeakPasswords[strings.ToLower(password)]
	return found
}

// isSingleRepeatedRune reports whether password is the same rune repeated
// for its entire length ("aaaaaaaa", "11111111", "        ") — the
// simplest possible predictable pattern, and one no blocklist entry could
// enumerate in advance since it exists for every rune, not just ASCII.
func isSingleRepeatedRune(password string) bool {
	runes := []rune(password)
	for i := 1; i < len(runes); i++ {
		if runes[i] != runes[0] {
			return false
		}
	}
	return true
}

// isSequentialRun reports whether password is a strictly ascending or
// strictly descending run of consecutive code points — "12345678",
// "abcdefgh", "87654321", "hgfedcba". This is a generic, alphabet-agnostic
// check rather than a hardcoded keyboard-layout table: it costs nothing to
// evaluate and catches the entire family of "just typed the next N
// characters" passwords without needing to enumerate them individually.
func isSequentialRun(password string) bool {
	runes := []rune(password)
	if len(runes) < 2 {
		return false
	}
	ascending, descending := true, true
	for i := 1; i < len(runes); i++ {
		delta := runes[i] - runes[i-1]
		if delta != 1 {
			ascending = false
		}
		if delta != -1 {
			descending = false
		}
	}
	return ascending || descending
}

File: web/src/features/auth/passwordStrength.ts
// Mirrors internal/user/password_strength.go's checks exactly — issue
// #219's own framing is "espelha a regra do servidor e nunca inventa uma
// mais frouxa" (mirrors the server's rule, never invents a looser one).
// This project has no cross-language codegen for a shared rule like
// this one, so the two files are kept in sync by hand: if the Go list or
// checks change, this file changes with them in the same commit — see
// docs/DECISIONS.md § "Validação de senha forte".

// Same ~150 entries as commonWeakPasswords in password_strength.go,
// lowercase. "password123" is deliberately absent there for the same
// reason it would be absent here: it is this project's own established
// demo/test password (registerAndLogin in web/e2e/fixtures.ts and every
// RegisterPage test use "correct horse battery staple" instead, so this
// omission has no effect on this app's own test fixtures either way).
const COMMON_WEAK_PASSWORDS = new Set([
  '123456', 'password', '12345678', 'qwerty', '123456789', '12345',
  '1234', '111111', '1234567', 'dragon', '123123', 'baseball',
  'iloveyou', 'trustno1', '1234567890', 'sunshine', 'master', '123321',
  'letmein', '696969', 'shadow', 'michael', 'superman', '1qaz2wsx',
  '7777777', '121212', '000000', 'qazwsx', '123qwe', 'killer',
  'jordan', 'jennifer', 'hunter', 'buster', 'soccer', 'harley',
  'batman', 'andrew', 'tigger', 'charlie', 'robert', 'thomas',
  'hockey', 'ranger', 'daniel', 'starwars', '112233', 'george',
  'computer', 'michelle', 'jessica', 'pepper', '1111', 'zxcvbn',
  '555555', '11111111', '131313', 'freedom', '777777', 'pass',
  'maggie', '159753', 'aaaaaa', 'ginger', 'princess', 'joshua',
  'cheese', 'amanda', 'summer', 'ashley', '6969', 'nicole',
  'chelsea', 'biteme', 'matthew', 'access', 'yankees', '987654321',
  'dallas', 'austin', 'thunder', 'taylor', 'matrix', 'mustang',
  'ninja', 'welcome', 'welcome1', 'admin', 'administrator', 'admin123',
  'root', 'toor', 'changeme', 'letmein1', 'password1',
  'passw0rd', 'p@ssw0rd', 'passw0rd!', 'password1!', 'qwerty123',
  'qwertyuiop', '1q2w3e4r', '1q2w3e4r5t', 'zaq1zaq1', 'aa123456',
  'asdfghjkl', 'abc12345', 'abcd1234', '1qaz2wsx3edc', 'iloveyou1',
  'whatever', 'trustme', 'letmein123', 'changeme123', 'temppass',
  'temp123', 'guest', 'guest123', 'test123', 'testing', 'default',
  'default123', 'incorrect', 'wordpass', 'mypassword', 'mypassword1',
  'hello123', 'hello1', 'welcome123', 'abc123', 'monkey', 'football',
  'jesus', 'flower', 'hottie', 'loveme', 'biteme1', 'sunshine1',
])

/** Mirrors isCommonWeakPassword in password_strength.go. */
export function isCommonWeakPassword(password: string): boolean {
  return COMMON_WEAK_PASSWORDS.has(password.toLowerCase())
}

/**
 * Mirrors isSingleRepeatedRune in password_strength.go. Array.from
 * splits by Unicode code point (not UTF-16 code unit), the same
 * granularity Go's []rune conversion gives the server-side check.
 */
export function isSingleRepeatedRune(password: string): boolean {
  const chars = Array.from(password)
  return chars.every((c) => c === chars[0])
}

/** Mirrors isSequentialRun in password_strength.go. */
export function isSequentialRun(password: string): boolean {
  const codePoints = Array.from(password).map((c) => c.codePointAt(0)!)
  if (codePoints.length < 2) return false
  let ascending = true
  let descending = true
  for (let i = 1; i < codePoints.length; i++) {
    // Non-null: i ranges over [1, codePoints.length), so both indices
    // are always in bounds — noUncheckedIndexedAccess can't infer that
    // from a numeric loop bound the way it can from destructuring.
    const delta = codePoints[i]! - codePoints[i - 1]!
    if (delta !== 1) ascending = false
    if (delta !== -1) descending = false
  }
  return ascending || descending
}

export interface PasswordRequirement {
  id: string
  label: string
  met: (password: string) => boolean
}

/**
 * The requirements PasswordRequirements.tsx renders as a live checklist.
 * Deliberately a checklist against the server's own predicates, not a
 * weak/fair/strong score — a score is exactly the kind of thing that
 * could tell a user their password is fine when the server would still
 * reject it. minPasswordLen/maxPasswordLen are mirrored as 8/72 directly
 * (not imported — there is nothing to import from across a language
 * boundary); the 72-character ceiling isn't listed here since it isn't a
 * strength concern, the same reasoning validatePassword's Go doc comment
 * gives.
 */
export const PASSWORD_REQUIREMENTS: PasswordRequirement[] = [
  { id: 'length', label: 'At least 8 characters', met: (p) => p.length >= 8 },
  {
    id: 'not-common',
    label: 'Not a commonly used password',
    met: (p) => !isCommonWeakPassword(p),
  },
  {
    id: 'not-predictable',
    label: 'Not a repeated or sequential pattern',
    met: (p) => !isSingleRepeatedRune(p) && !isSequentialRun(p),
  },
]
